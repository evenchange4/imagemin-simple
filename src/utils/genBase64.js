// @flow
import fsJson from 'fs-extra/lib/json';
import * as R from 'ramda';
import globby from 'globby';
import { DEFAULT_PATTERN } from './constants';
import { type Base64Arguments, type Base64Image } from './type.flow';
import log from './log';
import readBase64Image from './readBase64Image';

const genBase64 = async ({ pattern, outputFile }: Base64Arguments) => {
  try {
    const imagePaths = await globby([pattern, ...DEFAULT_PATTERN]);
    if (imagePaths.length > 0) log(JSON.stringify(imagePaths, null, 2));

    // Note: 1. gen base64 images and metadata
    const base64Images: Array<Base64Image> = await Promise.all(
      imagePaths.map(readBase64Image),
    );

    // Note: 2. output json format $FlowFixMe
    const results: { [string]: Base64Image } = R.indexBy(
      R.prop('imagePath'),
      base64Images,
    );
    fsJson.writeJson(outputFile, results);
    log(
      `> Generate ${imagePaths.length} images in base64 format successfully.`,
    );
  } catch (error) {
    throw error;
  }
};

export default genBase64;
