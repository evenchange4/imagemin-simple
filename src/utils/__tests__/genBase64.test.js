// @flow
import path from 'path';
import fsJson from 'fs-extra/lib/json';
import genBase64 from '../genBase64';

jest.mock('../readBase64Image', () => async () => ({
  base64: 'mock-readBase64Image-return-base64',
  imagePath: 'mock-readBase64Image-return-imagePath',
}));

it('should return base64 json correctly', async () => {
  const pattern = path.resolve(process.cwd(), '**/*.{jpg,png,JPG,PNG}');
  const outputFile = path.resolve(
    process.cwd(),
    'node_modules/.cache/geBase64.test.json',
  );
  const spy = jest.spyOn(fsJson, 'writeJson');

  await genBase64({ pattern, outputFile });
  expect(spy.mock.calls).toMatchSnapshot();
});
