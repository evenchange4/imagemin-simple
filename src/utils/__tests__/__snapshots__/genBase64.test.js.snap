// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return base64 json correctly 1`] = `
Array [
  Array [
    "/Users/mtk07866/code/GitHub/imagemin-simple/node_modules/.cache/base64.json",
    "{
  \\"mock-readBase64Image-return-imagePath\\": {
    \\"base64\\": \\"mock-readBase64Image-return-base64\\",
    \\"imagePath\\": \\"mock-readBase64Image-return-imagePath\\"
  }
}",
  ],
]
`;
